radix-engine-stores/src/hash_tree.rs

```diff
@@ -0,0 +1,256 @@
+            self.store
+                .insert_node(&TreeNodeKey::from(key), TreeNode::from(key, node));
+        }
+        for stale_node in update_result.stale_node_index_batch.iter().flatten() {
+            let key = TreeNodeKey::from(&stale_node.node_key);
+            self.store.record_stale_node(&key);
+        }
+        self.current_version = next_version;
+    }
+
+    /// Returns the hash of a state at version [`HashTree::current_version`].
+    pub fn get_current_root_hash(&self) -> Hash {
+        JellyfishMerkleTree::new(self.store)
+            .get_root_hash(self.current_version)
+            .expect("error while reading root hash")
+    }
+}
+
+impl From<&[u8]> for Nibs {
+    fn from(bytes: &[u8]) -> Self {
+        Nibs(
+            bytes
+                .iter()
+                .map(|byte| [Nib(byte >> 4), Nib(byte & 15)])
+                .flatten()
+                .collect(),
+        )
+    }
+}
+
+impl From<&Nibs> for Vec<u8> {
+    fn from(nibs: &Nibs) -> Self {
+        nibs.0
+            .chunks(2)
+            .map(|chunk| chunk[0].0 << 4 | chunk[1].0)
+            .collect::<Vec<u8>>()
+    }
+}
+
+impl TreeInternalNode {
+    fn from(internal_node: &InternalNode) -> Self {
+        let children = internal_node
+            .children_sorted()
+            .map(|(nibble, child)| TreeChildEntry {
+                nib: Nib(nibble.clone().into()),
+                version: child.version,
+                hash: child.hash,
+                is_leaf: child.is_leaf(),
+            })
+            .collect::<Vec<TreeChildEntry>>();
+        TreeInternalNode { children }
+    }
+}
+
+impl TreeLeafNode {
+    fn from(key: &NodeKey, leaf_node: &LeafNode<SubstateId>) -> Self {
+        TreeLeafNode {
+            nib_suffix: Nibs::from(leaf_node.account_key().as_ref())
+                .skip(key.nibble_path().num_nibbles()),
+            substate_id: leaf_node.value_index().0.clone(),
+            value_hash: leaf_node.value_hash(),
+        }
+    }
+}
+
+impl TreeNode {
+    fn from(key: &NodeKey, node: &Node<SubstateId>) -> Self {
+        match node {
+            Node::Internal(internal_node) => {
+                TreeNode::Internal(TreeInternalNode::from(internal_node))
+            }
+            Node::Leaf(leaf_node) => TreeNode::Leaf(TreeLeafNode::from(key, leaf_node)),
+            Node::Null => TreeNode::Null,
+        }
+    }
+}
+
+impl InternalNode {
+    fn from(internal_node: &TreeInternalNode) -> Self {
+        let child_map: HashMap<Nibble, Child> = internal_node
+            .children
+            .iter()
+            .map(|child_meta| {
+                let child: Child = Child::new(
+                    child_meta.hash,
+                    child_meta.version,
+                    if child_meta.is_leaf {
+                        NodeType::Leaf
+                    } else {
+                        NodeType::Internal { leaf_count: 0 }
```

Comment:
The `leaf_count: 0` [deliberately](https://github.com/radixdlt/radixdlt-scrypto/pull/763#issue-1551007534) skips this "statistics-only" feature from the underlying JMT impl for now.